### PR TITLE
HL-161

### PR DESCRIPTION
**Pull request by bazmatic**
https://dltx.atlassian.net/browse/HL-161

- Add refund() method to market contracts
- Add scratchAndRefund() method to market contracts
- Tests

### CODE REVIEW

File: contracts/Market.sol [hunk 1]
```diff
@@ -356,7 +358,44 @@ contract Market is IMarket, Ownable, ERC721 {
...
+	 * @param signature Signature from market owner
+	 */
+	function refund(uint64 index, SignatureLib.Signature calldata signature) external {
+		bytes16 REFUND_CMD = 0x726566756e6400000000000000000000; //Bytes for "refund"
+		bytes32 messageHash = keccak256(
+			abi.encodePacked(REFUND_CMD, address(this), index)
+		);
+
+		require(
+			isValidSignature(messageHash, signature) == true,
```
Comment (brettvniekerk) [inline contracts/Market.sol:376]: can just do `isValidSignature` no? also i feel like these checks should be in a modifier if they aren't already

File: contracts/Market.sol [hunk 2]
```diff
@@ -356,7 +358,44 @@ contract Market is IMarket, Ownable, ERC721 {
...
+
+	/* 
+	 * @dev Reverse a bet. Return the stake to the bettor and the loan to the vault
+	 * @param index The index of the bet
+	 * @param signature Signature from market owner
+	 */
+	function refund(uint64 index, SignatureLib.Signature calldata signature) external {
+		bytes16 REFUND_CMD = 0x726566756e6400000000000000000000; //Bytes for "refund"
+		bytes32 messageHash = keccak256(
+			abi.encodePacked(REFUND_CMD, address(this), index)
```
Comment (brettvniekerk) [inline contracts/Market.sol:372]: not sure what the point of passing `REFUND_CMD` is - wouldnt the `signature` contain the index of the bet that is to be refunded, and then the sig validity is checked?

File: contracts/Market.sol [hunk 3]
```diff
@@ -356,7 +358,44 @@ contract Market is IMarket, Ownable, ERC721 {
...
+
+		bet.settled = true;
+		uint256 loan = _bets[index].payout - _bets[index].amount;
+		_totalExposure -= loan;
+		_totalInPlay -= _bets[index].amount;
+		_inplayCount --;
+		
+		IERC20(_vault.asset()).transfer(ownerOf(index), bet.amount);
+		IERC20(_vault.asset()).transfer(address(_vault), loan);
+		emit Repaid(_vault.asset(), loan);
```
Comment (brettvniekerk) [inline contracts/Market.sol:394]: afaik dont events come before state changes?

File: contracts/Market.sol [hunk 4]
```diff
@@ -356,7 +358,44 @@ contract Market is IMarket, Ownable, ERC721 {
...
+			"refund: Invalid signature"
+		);
+		_refund(index);
+	}
+
+	function _refund(uint64 index) internal virtual {
+		Bet memory bet = _bets[index];
+		require(bet.settled == false, "refund: Bet has already settled");
+
+		bet.settled = true;
```
Comment (brettvniekerk) [inline contracts/Market.sol:386]: maybe we should have an `enum BetStatus` that can be `INPLAY`, `SETTLED`, or `REFUNDED`?

File: tests/utils.ts [hunk 5]
```diff
@@ -120,6 +129,18 @@ export const makeSetResultMessage = (
 	return message;
 };
 
+export const makeRefundMessage = (
+	marketAddress: string,
+	betIndex: BigNumberish
+): string => {
+	const b16Refund = formatBytes16String("refund");
```
Comment (brettvniekerk) [inline tests/utils.ts:136]: as i said earlier this seems unnecessary to me

File: contracts/Market.sol (hunk 3 above)
Comment (bitcoinbrisbane) [reply contracts/Market.sol:394]: No, events come after so you know the above line has excited.  Check the examples here: https://consensys.net/blog/developers/solidity-best-practices-for-smart-contract-security/

image

File: contracts/MarketCollateralised.sol [hunk 6]
```diff
@@ -46,14 +46,25 @@ abstract contract MarketCollateralised is Market {
...
-		bytes16 propositionId,
-		bytes16 marketId
-	) internal view returns (bool) {
-		return
-			keccak256(abi.encodePacked(propositionId)) ==
-			keccak256(abi.encodePacked(_mostExpensivePropositionId[marketId]));
-	}*/
+	function _refund(uint64 index) internal override {
+		Bet memory bet = _bets[index];
+		require(bet.settled == false, "refund: Bet has already settled");
```
Comment (bitcoinbrisbane) [inline contracts/MarketCollateralised.sol:51]:
```suggestion
		require(bet.settled == false, "_refund: Bet has already settled");
```

File: contracts/Market.sol (hunk 2 above)
Comment (bazmatic) [reply contracts/Market.sol:372]: > Using an instruction string and contract address in signed message. Instruction string refund makes the intent of the signed message clear. 

If you're signing something that is just a number, it's not clear what the intent of the signature is. Including a parameter that provides the intent means that the signatory can gain some clarity on what the signature is for.

File: contracts/Market.sol (hunk 4 above)
Comment (bazmatic) [reply contracts/Market.sol:386]: Interesting idea

File: tests/utils.ts (hunk 5 above)
Comment (bazmatic) [reply tests/utils.ts:136]: See above

File: contracts/Market.sol (hunk 1 above)
Comment (bazmatic) [reply contracts/Market.sol:376]: No because you have to generate the messageHash first.

File: contracts/Market.sol (hunk 4 above)
Comment (bitcoinbrisbane) [reply contracts/Market.sol:386]: This has been discussed, it doesn't make sense.

1. The bet is always in play
2. Or it's paid out.

File: tests/market.collateralised.test.ts [hunk 7]
```diff
@@ -718,4 +719,19 @@ describe("Collateralised Market: play through", function () {
 			"Total collateral should go down by the amount taken from spare collateral"
 		).to.equal(ethers.utils.parseUnits("4", tokenDecimals));
 	});
+
```
Comment (bazmatic) [inline tests/market.collateralised.test.ts:722]: Let's delete this fella

File: tests/market.test.ts [hunk 8]
```diff
@@ -23,7 +23,7 @@ import { formatBytes16String } from "../scripts/utils";
 
 chai.use(solidity);
 
-describe("Market", () => {
+describe.only("Market", () => {
```
Comment (bazmatic) [inline tests/market.test.ts:26]:
```suggestion
describe("Market", () => {
```

**Pull request merged**